src/apiutil.h

```diff
@@ -308,3 +316,208 @@ bool hasInsufficientMaterial(Color c, const Position& pos) {
+
+  int nbProcessedRanks = 1;  // start with 1 and increment after delimeter
+  int curRankWidth = 0;
+  char prevChar = '?';
+  for (char c : fenParts[0]) {
+    if (isdigit(c)) {
+      curRankWidth += c - '0';
+      // if previous digit exists we have a dual digit
+      if (isdigit(prevChar))
+        curRankWidth += 9;
+    }
+    else if (c == '/') {
+      ++nbProcessedRanks;
+      if (curRankWidth != nbFiles) {
+        std::cerr << "curRankWidth != nbFiles: " << curRankWidth << " != " << nbFiles << std::endl;
+        return FEN_FILE_SIZE_MISMATCH;
+     }
+      curRankWidth = 0;
+    }
+    else if (c == '+') {
+        // ignore '+' symbols in fen (e.g. kyotoshogi_variant)
+    }
+    else  // normal piece
+      ++curRankWidth;
+    prevChar = c;
+  }
+  if (v->pieceDrops) { // pockets can either be defined by [] or /
+    if (nbProcessedRanks != nbRanks && nbProcessedRanks-1 != nbRanks) {
+      std::cerr << "Invalid number of ranks. Expected: " << nbRanks << " Actual: " << nbProcessedRanks << std::endl;
+      return FEN_INVALID_NUMBER_OF_RANKS;
+    }
+  }
+  else {
+    if (nbProcessedRanks != nbRanks) {
+      std::cerr << "Invalid number of ranks. Expected: " << nbRanks << " Actual: " << nbProcessedRanks << std::endl;
+      return FEN_INVALID_NUMBER_OF_RANKS;
+    }
+  }
+
+  // check for pocket
+  if (v->pieceDrops) {
+    if (nbProcessedRanks-1 == nbRanks) {
+      // look for last '/'
+      for (auto it = fenParts[0].rbegin(); it != fenParts[0].rend(); ++it) {
+        if (*it == '/')
+          break;
+        if (*it != '-') {
+          if (v->pieceToChar.find(*it) == std::string::npos) {
+            std::cerr << "Invalid pocket piece: '" << *it << "'." << std::endl;
+            return FEN_INVALID_POCKET_INFO;
+          }
+        }
+      }
+    }
+    else {  // pocket is defined as [ and ]
+      if (*(fenParts[0].end()-1) != ']') {
+        std::cerr << "Pocket specification does not end with ']'." << std::endl;
+        return FEN_INVALID_POCKET_INFO;
+      }
+      // look for first '['
+      for (auto it = fenParts[0].rbegin()+1; it != fenParts[0].rend(); ++it) {
+        if (*it == '[')
+          break;
+        if (*it != '-')
+          if (v->pieceToChar.find(*it) == std::string::npos) {
+              std::cerr << "Invalid pocket piece: '" << *it << "'." << std::endl;
+            return FEN_INVALID_POCKET_INFO;
+          }
+      }
+    }
+  }
+
+  // 2) Part
+  // check side to move char
+  if (fenParts[1][0] != 'w' && fenParts[1][0] != 'b') {
+    std::cerr << "Invalid side to move specification: '" << fenParts[1][0] << "'." << std::endl;
+    return FEN_INVALID_SIDE_TO_MOVE;
+  }
+
+  // 3) Part
+  // check castling rights (exclude seirawan_variant because it has different castling specification)
+  if (v->castling && !v->seirawanGating) {
+    for (char c : fenParts[2])
+      if (c != '-')
+        if (c != 'k' && c != 'K' && c != 'Q' && c != 'q') {
```

Comment:
Chess960 needs to be considered here, since in shuffling/960 variants the castling rights can be any letters that denote valid files. Since we can not know here if `UCI_Chess960` is enabled, one perhaps always has to assume it might be the case.

Reply:
I think we can check for `UCI_Chess960` using `v->chess960`.
Therefore, we can relax the castling specification in this case to allow any non-digit character.

Reply:
No, this unfortunately is not possible. The "chess960" attribute in a variant only enforces chess960, even if the UCI option is disabled, but for variants where it is false the UCI option could still be enabled. This logic is to ensure compatibility with the normal UCI option and at the same time allowing to configure variants like caparandom where the UCI option should be ignored. Also see
https://github.com/ianfab/Fairy-Stockfish/blob/369c55f9f2dce34b9cb65ba78ef1649543d11f59/src/position.cpp#L457

Reply:
I'll relax the castling specification condition to use `isalpha(c)` instead.